x-pack/plugins/fleet/server/routes/package_policy/handlers.ts

```diff
@@ -43,16 +43,68 @@ import { simplifiedPackagePolicytoNewPackagePolicy } from '../../../common/servi
 
 import type { SimplifiedPackagePolicy } from '../../../common/services/simplified_package_policy_helper';
 
-export const getPackagePoliciesHandler: RequestHandler<
+export const getPackagePoliciesHandler: FleetRequestHandler<
   undefined,
   TypeOf<typeof GetPackagePoliciesRequestSchema.query>
 > = async (context, request, response) => {
-  const soClient = (await context.core).savedObjects.client;
+  const { client: soClient, limitedToPackages } = await (await context.fleet).getSoClient();
+
   try {
     const { items, total, page, perPage } = await packagePolicyService.list(
       soClient,
       request.query
     );
+
+    // specific to package-level RBAC
+    if (limitedToPackages && limitedToPackages.length) {
+      // { packageName: boolean } record of allowed package names
+      // include policy with no package name
+      const packagePolicyNames = items.reduce<Record<string, boolean>>((acc, item) => {
+        const pkgName = item?.package?.name ?? 'empty-name';
+        if (pkgName) {
+          acc[pkgName] = limitedToPackages.includes(pkgName);
+        }
+        return acc;
+      }, {});
+
+      if (Object.values(packagePolicyNames).every((value) => !value)) {
+        // when there are not any allowed policies
+        return response.forbidden({
+          body: {
+            message: `Data for package names ${items
+              .map((item) => item.package?.name)
+              .join(', ')} is not authorized.`,
+          },
+        });
+      }
+
+      // list allowed package policy items or policies with no package name
+      return response.ok({
+        body: {
+          items: items.filter((item) => {
+            const pkgName = item?.package?.name;
+            // include policies with allowed package names
+            const allowedPackages = Object.entries(packagePolicyNames).reduce<string[]>(
+              (acc, curr) => {
+                if (curr[1]) acc.push(curr[0]);
+                return acc;
+              },
+              []
+            );
+            if (pkgName) {
+              return allowedPackages.includes(pkgName);
+            }
+            // or policies with no package name
+            return item;
+          }),
+          total,
+          page,
+          perPage,
+        },
+      });
```

Comment:
I don't understand what this code is attempting to do?

Reply:
I wasn't sure about this actually, but here it is filtering put policy items based on allowed package names. In case there are policies without a package name (not sure if there ever is such a case but at leat the type says it's optional) it allows it in the list.

Reply:
What you mean "put"? this is the `get list` api.

I'm going to remove this. Also - for now, although the Type definition indicates that `package` is optional, I don't know of a use case where it might be undefined - perhaps when Agent is used in standalone mode?

For now, we I propose we go with: if `package` is not defined, then we will error out. This seems like the right thing to do being that we are adding access based on `package privileges`, so a package is a requirement. 

Note: I'm also changing a bit your logic around validation and error out. We should exit the moment we encounter an invalid `package.name` and not attempt to loop through all of the data. That because the API's in fleet do not bound the `perPage` value, so an API request could theoretically set this to a very large value and I don't think we should spend time collecting all of the invalid package name. New implementation will just report the first package name that is not authorized and also provide the list of package name that IS authorized.